src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -738,99 +640,42 @@ void DiscoveryDataBase::process_dispose_participant(
     }
 }
 
-void DiscoveryDataBase::process_dispose_writer(
-        eprosima::fastrtps::rtps::CacheChange_t* ch,
-        const std::string& topic_name)
+void DiscoveryDataBase::process_dispose_writer_(
+        eprosima::fastrtps::rtps::CacheChange_t* ch)
 {
     const eprosima::fastrtps::rtps::GUID_t& writer_guid = guid_from_change(ch);
 
     // Change DATA(w) with DATA(Uw)
     std::map<eprosima::fastrtps::rtps::GUID_t, DiscoveryEndpointInfo>::iterator wit = writers_.find(writer_guid);
     if (wit != writers_.end())
     {
-        changes_to_release_.push_back(wit->second.update_and_unmatch(ch));
+        update_change_and_unmatch_(ch, wit->second);
     }
 
-    // Update own entry participants_::writers
-    std::map<eprosima::fastrtps::rtps::GuidPrefix_t, DiscoveryParticipantInfo>::iterator pit =
-            participants_.find(writer_guid.guidPrefix);
-    if (pit != participants_.end())
-    {
-        pit->second.remove_writer(writer_guid);
-    }
-
-    // Update own entry writers_by_topic_
-    std::map<std::string, std::vector<eprosima::fastrtps::rtps::GUID_t>>::iterator tit =
-            writers_by_topic_.find(topic_name);
-    if (tit != writers_by_topic_.end())
-    {
-        for (std::vector<eprosima::fastrtps::rtps::GUID_t>::iterator writer_it = tit->second.begin();
-                writer_it != tit->second.end();
-                ++writer_it)
-        {
-            if (*writer_it == writer_guid)
-            {
-                tit->second.erase(writer_it);
-                break;
-            }
-        }
-
-        if (tit->second.empty())
-        {
-            writers_by_topic_.erase(tit);
-        }
-    }
+    // Update writer matches
+    unmatch_writer_(writer_guid);
```

Comment:
This should go inside the previous `if` statement. Since first we process the PDP queue and then the EDP, it could be that we have already processed the DATA(Up) and have already unmatched the writer by the time we process the DATA(Uw). In that case we actually do not need to propagate the DATA(Uw), the DATA(Up) will be enough for everyone.

Reply:
Moreover, the `writers_.find(...)`  is repeated at the beginning of the `unmatch_writer()` with a `return` if it is not found. The way @EduPonz says, we avoid the second find.

Reply:
:+1: